src/core/namespace.c

```diff
@@ -1639,6 +1700,111 @@ static int setup_one_tmp_dir(const char *id, const char *prefix, char **path, ch
         return 0;
 }
 
+static int append_extension_images(
+                MountEntry **p,
+                const char *root,
+                const char *propagate_dir,
+                char **hierarchies,
+                const MountImage *mount_images,
+                size_t n) {
+
+        _cleanup_free_ char *mount_point_parent = NULL, *propagate_parent = NULL;
+        _cleanup_strv_free_ char **overlays = NULL;
+        char **hierarchy;
+        int r;
+
+        assert(p);
+
+        if (n == 0)
+                return 0;
+
+        /* On systemd --user instance we don't have a propagate dir. */
+        if (!propagate_dir)
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to create temp directory for overlays, no propagate directory available");
+
+        /* Prepare a list of overlays, that will have as each element a string suitable for being
+         * passed as a lowerdir= parameter, so start with the hierachy on the root.
+         * The overlays vector will have the same number of elements and will correspond to the
+         * hierarchies vector, so they can be iterated upon together. */
+        STRV_FOREACH(hierarchy, hierarchies) {
+                _cleanup_free_ char *prefixed_hierarchy = NULL;
+
+                prefixed_hierarchy = path_join(root, *hierarchy);
+                if (!prefixed_hierarchy)
+                        return -ENOMEM;
+
+                strv_consume(&overlays, TAKE_PTR(prefixed_hierarchy));
+        }
+
+        propagate_parent = dirname_malloc(propagate_dir);
```

Comment:
mm, what does this have to use with popagation? i don't get it? you could just mount this to `/run/systemd/unit-extension/{0,1,2,3,4,5}` or so, no? i.e. do it like for the root image, which we mount to `/run/systemd/unit-root`. Given that mount propagation for the whole of /run is turned off here, you can just take a dir of your own. (And i'd number the mounts, since after all the mount points to overmount have to be created and that way you don't need to clean them up really)

Reply:
Didn't want to add a new directory, and that one was already passed through as a unit-private dir. But changed to numbered mount points unbder unit-extension/ as suggested, as it is indeed much nicer.